### DISCUSSION TITLE
Custom aspect ratio for videos

### DISCUSSION THREAD

**Discussion by mhasanmeet** [elementor/elementor#23464] [Feature Request]
### Prerequisites

- [X] I have searched for similar features requests in both open and closed tickets and cannot find a duplicate.
- [X] The feature is still missing in the latest stable version of Elementor ( Elementor Pro )

### What problem is your feature request going to solve? Please describe.

When working with videos, the common aspect ratio that is given by elementor is not fulfilled my requirements, It needs custom aspect ratio input values. Something like this, 
![Screenshot_6](https://github.com/elementor/elementor/assets/24842666/38813a79-c7ea-4af2-b0f0-b781095f66f9)

### Describe the solution you'd like

It will be great if Elementor gives us freedom and let us choose our own aspect ratio for videos. It is preferable and can meet the requirements.

**Comment by namealreadytaken2** [comment]
+1

**Comment by da-ikuki** [comment]
+1
Any idea for css solution for the moment?
Thanks! ;)

**Comment by da-ikuki** [comment]
I've just found this :
**.elementor-wrapper {--video-aspect-ratio: 30.1 !important;
}**
Here, _30.1_ is my custom ratio. You can change it for your own needs.

**Warning :** this works for me, but there's an alert in Elementor custom CSS window : _"Expected RBRACE at line 1, col 21. Rule is empty."_... I'm not a CSS pro and I can't find where RBRACE is missing... Use this carefully and at your own risk. 😅 And please, don't hesitate to put the finishing touch to this CSS if you have any idea for improving.

Thanks!

**Comment by secondsky** [comment]
indeed this is an absolute must to implement.
I have no clue who at elementors team did this nonsense with fixed aspect ratios. What about video banners? or whatever else?
And the worst, it ignores custom CSS ID/classes and can only overridden if every single time the specific widget name is found through inspector. the above solution with wrapper can only be used if all the videos on a page should have the same aspect ratio, but if you use a video banner and the other widgets have a different aspect ratio, your screwed.

**Comment by zuzanu** [comment]
You can add this code to your functions.php file to implement a custom aspect ratio. Make sure to update the labels (e.g., 'label' => esc_html__( 'Breite', '4d-theme' )) and ensure that the text domain (4d-theme) matches your theme.

```
add_action('elementor/element/video/section_video_style/before_section_end',` function ($element, $args) {

    $element->remove_control( 'aspect_ratio' );

    // start injection of control before an other control
	$element->start_injection( [
		'at' => 'after',
		'of' => 'section_video_style',
	] );

	$element->add_responsive_control(
        'aspect_ratio',
        [
            'label' => esc_html__( 'Aspect Ratio', 'elementor' ),
            'type' => \Elementor\Controls_Manager::SELECT,
            'options' => [
                '1 / 1' => esc_html__( '1 / 1', 'elementor' ),
                '4 / 3' => esc_html__( '4 / 3', 'elementor' ),
                '3 / 4' => esc_html__( '3 / 4', 'elementor' ),
                '16 / 9' => esc_html__( '16 / 9', 'elementor' ),
                '16 / 10' => esc_html__( '16 / 10', 'elementor' ),
                '21 / 9' => esc_html__( '21 / 9', 'elementor' ),
                '' => esc_html__( 'Custom', 'elementor' ),
            ],
            'default' => '16 / 9',
            'selectors' => [
                '{{WRAPPER}} .elementor-wrapper' => '--video-aspect-ratio: {{VALUE}}'
            ],
        ]
    );

    $element->add_responsive_control(
        'aspect_ratio_width',
        [
            'label' => esc_html__( 'Breite', '4d-theme' ),
            'type' => \Elementor\Controls_Manager::NUMBER,
            'min' => 1,
	    'max' => 500,
	    'step' => 1,
            'default' => 16,
            'condition' => [
                'aspect_ratio' => '',
            ],
        ]
    );

    $element->add_responsive_control(
        'aspect_ratio_height',
        [
            'label' => esc_html__( 'Höhe', '4d-theme' ),
            'type' => \Elementor\Controls_Manager::NUMBER,
            'min' => 1,
	    'max' => 500,
	    'step' => 1,
            'default' => 9,
            'selectors' => [
                '{{WRAPPER}} .elementor-wrapper' => 'aspect-ratio: {{aspect_ratio_width.VALUE}} / {{aspect_ratio_height.VALUE}};',
            ],
            'condition' => [
                'aspect_ratio' => '',
            ],
        ]
    );

    // end injection just after
   $element->end_injection();
}, 10, 2);
```

**Reply by Rgri66** [reply]
> You can add this code to your functions.php file to implement a custom aspect ratio. Make sure to update the labels (e.g., 'label' => esc_html__( 'Breite', '4d-theme' )) and ensure that the text domain (4d-theme) matches your theme.
> 
Great solution! Elementor should make this a standard feature!
